app/src/main/java/org/oppia/app/hintsandsolution/HintsViewModel.kt

```diff
@@ -36,36 +36,52 @@ class HintsViewModel @Inject constructor() : HintsAndSolutionItemViewModel() {
       if (itemList.isEmpty()) {
-        val hintsAndSolutionViewModel = HintsViewModel()
-        hintsAndSolutionViewModel.title.set(hintList[index].hintContent.contentId)
-        hintsAndSolutionViewModel.hintsAndSolutionSummary.set(hintList[index].hintContent.html)
-        hintsAndSolutionViewModel.isHintRevealed.set(hintList[index].hintIsRevealed)
-        itemList.add(hintsAndSolutionViewModel as HintsAndSolutionItemViewModel)
-      } else if (index == itemList.size) {
-        val isLastHintRevealed = (itemList[index - 1] as HintsViewModel).isHintRevealed.get()!!
-        if (isLastHintRevealed) {
-          val hintsAndSolutionViewModel = HintsViewModel()
-          hintsAndSolutionViewModel.title.set(hintList[index].hintContent.contentId)
-          hintsAndSolutionViewModel.hintsAndSolutionSummary.set(hintList[index].hintContent.html)
-          hintsAndSolutionViewModel.isHintRevealed.set(hintList[index].hintIsRevealed)
-          itemList.add(hintsAndSolutionViewModel as HintsAndSolutionItemViewModel)
+        addHintToList(hintList[index])
+      } else {
+        val isLastHintRevealed =
+          (itemList[itemList.size - 2] as HintsViewModel).isHintRevealed.get()!!
+        if (isLastHintRevealed && index <= newAvailableHintIndex.get()!! / 2) {
+          addHintToList(hintList[index])
+        } else {
+          break
         }
       }
     }
-
-    val isLastHintRevealed = (itemList[itemList.size - 1] as HintsViewModel).isHintRevealed.get()!!
-    if (solution.hasExplanation() && hintList.size == itemList.size && isLastHintRevealed) {
-      val solutionViewModel = SolutionViewModel()
-      solutionViewModel.title.set(solution.explanation.contentId)
-      solutionViewModel.correctAnswer.set(solution.correctAnswer.correctAnswer)
-      solutionViewModel.numerator.set(solution.correctAnswer.numerator)
-      solutionViewModel.denominator.set(solution.correctAnswer.denominator)
-      solutionViewModel.wholeNumber.set(solution.correctAnswer.wholeNumber)
-      solutionViewModel.isNegative.set(solution.correctAnswer.isNegative)
-      solutionViewModel.solutionSummary.set(solution.explanation.html)
-      solutionViewModel.isSolutionRevealed.set(solution.solutionIsRevealed)
-      itemList.add(solutionViewModel)
+    val isLastHintRevealed = (itemList[itemList.size - 2] as HintsViewModel).isHintRevealed.get()!!
+    if (solution.hasExplanation() &&
+      hintList.size * 2 == itemList.size &&
+      isLastHintRevealed &&
+      allHintsExhausted.get()!!
+    ) {
+      addSolutionToList(solution)
     }
     return itemList
   }
+
+  private fun addHintToList(hint: Hint) {
+    val hintsAndSolutionViewModel = HintsViewModel()
+    hintsAndSolutionViewModel.title.set(hint.hintContent.contentId)
+    hintsAndSolutionViewModel.hintsAndSolutionSummary.set(hint.hintContent.html)
+    hintsAndSolutionViewModel.isHintRevealed.set(hint.hintIsRevealed)
+    itemList.add(hintsAndSolutionViewModel as HintsAndSolutionItemViewModel)
```

Comment:
This cast is a bit weird. Why is this needed?

Reply:
Removed this cast.
Also, renamed `hintsAndSolutionViewModel` to `hintsViewModel`